types/styled-components/index.d.ts

```diff
@@ -436,4 +436,29 @@ export class StyleSheetManager extends React.Component<
     StyleSheetManagerProps
 > {}
 
+export type CSSIntrinsicAttributeType =
+    | string
+    | CSSObject
+    | FlattenSimpleInterpolation
+    // Sad, but because this is global, there is no way to override it with the ThemedStyledComponentsModule
+    // Only augmenting DefaultTheme will work for inline css prop
+    | FlattenInterpolation<ThemeProps<AnyIfEmpty<DefaultTheme>>>;
+
 export default styled;
+
+declare module "react" {
+    interface Attributes {
+        // NOTE: unlike the plain javascript version, it is not possible to get access
+        // to the element's own attributes inside function interpolations.
+        // Only theme will be accessible, and only with the DefaultTheme due to the global
+        // nature of this declaration.
+        // If you are writing this inline you already have access to all the attributes anyway,
+        // no need for the extra indirection.
+        /**
+         * If present, this React element will be converted by
+         * `babel-plugin-styled-components` into a styled component
+         * with the given css as its styles.
+         */
+        css?: CSSIntrinsicAttributeType;
```

Comment:
My problem with this is that it virtually monkey-patches type of core react without having me sure I really wanted this. For users who don't use the plugin it will be very confusing that a new attribute is there.

I'm thinking if it can be moved to a separate file or package to be possible to indicate that intent.

Reply:
I could put it in the types of `/macro` instead; the plugin as loaded by babel-macros _will_ add support for `css`.

I can't think of any proper way of adding a type-only module that is importable on demand, though, other than by publishing something _outside_ DefinitelyTyped. While I can add something in DT, it'd have to be imported using `import {} from '...'` -- just directly using `import '...'` will cause runtime/bundling errors.

Reply:
I wonder what happens if I add a `css?: never` here with a JSDoc comment telling you how to load the `css?: CSSIntrinsicAttributeType` augment 🤔

They'll probably conflict but it's worth a try...

Reply:
I refactored this and also tested the actual limitations of the plugin.